### PR TITLE
[RFC] Future structure

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

it is just PoC, and it follows the latest PyPA project structure - https://github.com/pypa/sampleproject
it is said that using src is better; it does not accidentally use modules that are not installed...

so with this new structure, we would have
```
examples/  # the actual pl_examples
src/
 L pytorch_lightning  # the actual package in root
test/
 |- legacy  # downloaded tests data from S3
 L  unittests  # actual folder with tests
```

**NOTE:**
- there is no change in user experience while installing this repo from the source
- there shall not be any breaking changes for ongoing/open PRs; if GitHub flags some collisions, they shall be quickly resolved in the command line with `git rebase master` as it uses here way resolver
- this will be merged as `rebase & merge` for traceability and simple resolving potential conflicts

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @carmocca @akihironitta @borda @tchaton @justusschock @awaelchli

### CODE REVIEW

File: .azure-pipelines/gpu-benchmark.yml [hunk 1]
```diff
@@ -26,16 +27,22 @@ jobs:
   - job: benchmarks
     timeoutInMinutes: "90"
     cancelTimeoutInMinutes: "2"
-    pool: azure-gpus-spot
+    pool: azure-jirka-spot
```
Comment (akihironitta) [inline .azure-pipelines/gpu-benchmark.yml:30]: Just leaving a reminder here that we need to revert the pool name before merge.

File: .gitignore [hunk 2]
```diff
@@ -8,13 +8,13 @@ lightning_logs/
 .vscode/
 
 # Documentations
-docs/source/api
+docs/source-PL/api
```
Comment (akihironitta) [inline .gitignore:11]: nit: I'm for `docs/source-pl` (in lowercase).

File: .azure-pipelines/gpu-benchmark.yml (hunk 1 above)
Comment (Borda) [reply .azure-pipelines/gpu-benchmark.yml:30]: this is temp solution to revert previous machines to unblock our PL CI

File: .gitignore (hunk 2 above)
Comment (Borda) [reply .gitignore:11]: well lets leave this for later :)

File: .azure-pipelines/gpu-tests.yml [hunk 3]
```diff
@@ -109,14 +114,15 @@ jobs:
 
     - script: |
         set -e
-        python -m pytest pl_examples -v --maxfail=2 --durations=0
-        bash pl_examples/run_examples.sh --trainer.accelerator=gpu --trainer.devices=1
-        bash pl_examples/run_examples.sh --trainer.accelerator=gpu --trainer.devices=2 --trainer.strategy=ddp
-        bash pl_examples/run_examples.sh --trainer.accelerator=gpu --trainer.devices=2 --trainer.strategy=ddp --trainer.precision=16
+        bash run_ddp_examples.sh
```
Comment (carmocca) [inline .azure-pipelines/gpu-tests.yml:117]: Let's keep the test here with the other standalone tests

File: .azure-pipelines/gpu-tests.yml [hunk 4]
```diff
@@ -109,14 +114,15 @@ jobs:
...
+        bash run_pl_examples.sh --trainer.accelerator=gpu --trainer.devices=1
```
Comment (carmocca) [inline .azure-pipelines/gpu-tests.yml:118]: This script should be examples/pytorch/run_examples.sh

File: .azure-pipelines/gpu-tests.yml [hunk 5]
```diff
@@ -56,8 +55,9 @@ jobs:
         python -c "fname = 'requirements/strategies.txt' ; lines = [line for line in open(fname).readlines() if 'horovod' not in line] ; open(fname, 'w').writelines(lines)"
         CUDA_VERSION_MM=$(python -c "import torch ; print(''.join(map(str, torch.version.cuda.split('.')[:2])))")
         pip install "bagua-cuda$CUDA_VERSION_MM>=0.9.0"
-        pip install . --requirement requirements/devel.txt
-        pip install . --requirement requirements/strategies.txt
+        pip install -e .
+        pip install --requirement requirements/devel.txt
+        pip install --requirement requirements/strategies.txt
```
Comment (carmocca) [inline .azure-pipelines/gpu-tests.yml:60]:
```suggestion
        pip install -e .[strategies]
        pip install --requirement requirements/devel.txt
```

File: .azure-pipelines/hpu-tests.yml [hunk 6]
```diff
@@ -33,25 +32,32 @@ jobs:
...
+          'unittests_pl/plugins/precision/hpu/ops_bf16.txt' --hmp-fp32 \
+          'unittests_pl/plugins/precision/hpu/ops_fp32.txt' --forked \
+          --junitxml=hpu1_precision_test-results.xml
+      workingDirectory: test
       displayName: 'HPU precision test'
 
     - bash: |
         export PYTHONPATH="${PYTHONPATH}:$(pwd)"
-        python "pl_examples/hpu_examples/simple_mnist/mnist.py"
+        python "pl_hpu/mnist_sample.py"
```
Comment (carmocca) [inline .azure-pipelines/hpu-tests.yml:59]: The structure here should be `examples/pytorch/hpu`

File: .azure-pipelines/ipu-tests.yml [hunk 7]
```diff
@@ -68,16 +68,23 @@ jobs:
...
 
     - bash: |
         source ${{ variables.poplar_sdk }}/poplar-ubuntu*/enable.sh
         source ${{ variables.poplar_sdk }}/popart-ubuntu*/enable.sh
-        export POPTORCH_WAIT_FOR_IPU=1
-        export PL_RUN_IPU_TESTS=1
-        python -m coverage run --source pytorch_lightning -m pytest tests -vv --junitxml=$(Build.StagingDirectory)/test-results.xml --durations=50
+        cd src
+        python -m pytest pytorch_lightning
+      displayName: 'DocTests'
```
Comment (carmocca) [inline .azure-pipelines/ipu-tests.yml:79]: I wouldn't add doctesting to these accelerator jobs as it complicates the pipeline for contributors

File: .github/CODEOWNERS [hunk 8]
```diff
@@ -16,48 +16,48 @@
 /docs/                                  @edenlightning @tchaton @borda @awaelchli @RobertLaurella
 /.github/*.md                           @edenlightning @williamfalcon @borda
 /.github/ISSUE_TEMPLATE/                @edenlightning @borda @tchaton
-/docs/source/conf.py                    @borda @awaelchli @carmocca
-/docs/source/index.rst                  @williamfalcon
-/docs/source/levels                     @williamfalcon @RobertLaurella
-/docs/source/expertise_levels           @williamfalcon @RobertLaurella
+/docs/source-PL/conf.py                    @borda @awaelchli @carmocca
```
Comment (carmocca) [inline .github/CODEOWNERS:19]: Why do we need this `-` separator instead of

`docs/pytorch_lightning/source/...`

**Comment by Borda** [comment]
> @Borda is there anything left to do here? Can we close this?

shall be fine, but would like to check it, run some diff that it is minimal...
